LiteFarmOrg/LiteFarm
# Automatically add JIRA ticket number to commit messages

**Description**

This is a little feature I worked on for myself to automate the process of adding a JIRA ticket number to commit messages. I was getting a bit tired of having to remember to add it everytime I make a commit, so I experimented and came up with this solution. I just wanted to share it here in case others may benefit from it as well and potentially merge it in the repository. 

**What it does**
Automatically adds the JIRA ticket number to the commit message on each commit.

**How it works**
Git natively provides several hooks that allow you to do something on certain actions. Like the pre-commit hook allows you to run some code before committing, there are several others documented [here](https://git-scm.com/docs/githooks). 

For this case I used the prepare-commit-msg hook that allows you to modify commit messages before a commit. Since LiteFarm is using husky, I set up the hook under .husky as mentioned in their docs. I used node.js because that's what I'm most familiar with but it can be done in other ways like using bash scripting.  

What the script does is checks for the presence of a Jira ticket number in the branch name, and if it exists, reads the commit message and adds it to the front.

**Tests**
I added tests for automatically testing a few different scenarios and edge cases. Initially I set it up without mocking - by creating a test branch, making test commits on that branch, then checking for the presence of ticket number in the commit message. But I found that it was slow (500ms - 1s) and prone to inconsistencies, so I ended up refactoring to use mocks. 

_Note:_ 
- The tests use built-in node test runner so nothing is required to be installed, but node.js version 18 or higher is needed. For best results, use version 20, as it has the stable release for the test runner.
- To run the tests, use command `node --test /pathToTestFile`

**How to test manually:**
1. Checkout this branch locally and run npm install or npm run prepare
2. Change the branch name so that a fake jira ticket number is appended to it. Like LF-XXXX branch-name.
3. Make a test commit without adding a ticket number to it.
4. Hopefully you should have the JIra ticket number automatically added to the message.

**Type of change**

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration


**Checklist:**
